test/Microsoft.Azure.WebJobs.Extensions.ServiceBus.Tests/ServiceBusSessionsEndToEndTests.cs

```diff
@@ -344,13 +349,78 @@ public async Task TestBatch_DataContractPoco()
             await TestMultiple<ServiceBusMultipleMessagesTestJob_BindToPocoArray>(true);
         }
 
-        private async Task TestMultiple<T>(bool isXml = false)
+        [Fact]
+        public async Task MessageDraining_QueueWithSessions()
+        {
+            await TestMultipleDrainMode<DrainModeValidationFunctions>(true);
+        }
+
+        [Fact]
+        public async Task MessageDraining_TopicWithSessions()
+        {
+            await TestMultipleDrainMode<DrainModeValidationFunctions>(false);
+        }
+
+        [Fact]
+        public async Task MessageDraining_QueueWithSessions_Batch()
+        {
+            await TestMultipleDrainMode<DrainModeValidationFunctions>(true);
+        }
+
+        [Fact]
+        public async Task MessageDraining_TopicWithSessions_Batch()
+        {
+            await TestMultipleDrainMode<DrainModeValidationFunctions>(false);
+        }
+
+        /*
+         * Helper functions
+         */
+
+        private async Task TestSingleDrainMode<T>(bool sendToQueue)
+        {
+            var host = ServiceBusSessionsTestHelper.CreateHost<DrainModeValidationFunctions>(_nameResolver);
+            await host.StartAsync();
+
+            _drainValidationPreDelay = new ManualResetEvent(initialState: false);
+            _drainValidationPostDelay = new ManualResetEvent(initialState: false);
+
+            if (sendToQueue)
+            {
+                await ServiceBusEndToEndTests.WriteQueueMessage(
+                    _connectionString, _queueName, "queue-message-draining-with-sessions-1", _drainModeSessionId);
+            }
+            else
+            {
+                await ServiceBusEndToEndTests.WriteTopicMessage(
+                    _connectionString, _topicName, "topic-message-draining-with-sessions-1", _drainModeSessionId);
+            }
+
+            // Wait to ensure function invocatoin has started before draining messages
+            Assert.True(_drainValidationPreDelay.WaitOne(SBTimeout));
+
+            // Start draining in-flight messages
+            var drainModeManager = host.Services.GetService<IDrainModeManager>();
+            await drainModeManager.EnableDrainModeAsync(CancellationToken.None);
```

Comment:
do we call `StopAsync` for the SB listener here?

Reply:
Yes.